scientific-computing-with-python/learn-special-methods-by-building-a-vector-space/vectors.py

```diff
@@ -0,0 +1,87 @@
+class Vector:
+    space = 'R3'
+
+    def __init__(self, x, y, z):
+        self.x = x
+        self.y = y
+        self.z = z
+        self.norm = (x**2 + y**2 + z**2)**0.5
+
+    def __str__(self):
+        return str((self.x, self.y, self.z))
+    
+    def __repr__(self):
+        return f'Vector({self.x}, {self.y}, {self.z})'
+
+    def __add__(self, other):        
+        x = self.x + other.x
+        y = self.y + other.y
+        z = self.z + other.z
+        v = Vector(x, y, z)
+        return v
```

Comment:
```suggestion
        if not isinstance(other, Vector):
            return NotImplemented
        x = self.x + other.x
        y = self.y + other.y
        z = self.z + other.z
        return Vector(x, y, z)
```
`Vector` class doesn't know how to add itself to the `other`'s class, however the class of `other` might know how to add itself to `Vector`. The `return NotImplemented` will allow python to automatically try that possibility.

Applies to other methods too.